pkg/cni/token_watch.go

```diff
@@ -0,0 +1,98 @@
+package cni
+
+import (
+	"encoding/base64"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/fsnotify/fsnotify.v1"
+	"k8s.io/client-go/rest"
+)
+
+var tokenPath string = "/var/run/secrets/kubernetes.io/serviceaccount/"
+
+func Run() {
+	// Create a watcher for file changes.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		panic(err)
+	}
+	defer watcher.Close()
+
+	// Start a go routine for handling events from the notifier.
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				// We've received a notification that the Kubernetes secrets files have changed.
+				// Update the kubeconfig file on disk to match.
+				logrus.WithField("event", event).Info("Notified of change to serviceaccount files")
+				cfg, err := rest.InClusterConfig()
+				if err != nil {
+					logrus.WithError(err).Error("Error generating kube config")
+					continue
+				}
+				err = rest.LoadTLSFiles(cfg)
+				if err != nil {
+					logrus.WithError(err).Error("Error loading TLS files")
+					continue
+				}
+				writeKubeconfig(cfg)
+
+			case err := <-watcher.Errors:
+				// We've received an error - log it out but don't exit.
+				logrus.WithError(err).Error("Error watching serviceaccount files")
+			}
+		}
+	}()
+
+	// Watch for changes to the serviceaccount token file.
+	for {
```

Comment:
Couldn't we just do this for before the go routine and do away with needing to run a go routine or reading from the stop below?

Reply:
Yeah I'll give that a try. This was done basically just by following the fsnotify examples :)